TopPlaces/TopPlaces/Controllers/DetailsPhotoViewController.m

```diff
@@ -0,0 +1,184 @@
+  
+  self.noImageLoadedView.hidden = YES;
+  [self.spinner startAnimating];
+
+  __weak DetailsPhotoViewController *weakSelf = self;
+  dispatch_queue_t fetchPhoto = dispatch_queue_create("picture of photo", NULL);
+  dispatch_async(fetchPhoto, ^(void){
+    //get the image
+    UIImage *img = [weakSelf.placesPhotosProvider downloadPhoto:weakSelf.photoInfo];
+    
+    //show it (on UI thread)
+    dispatch_async(dispatch_get_main_queue(), ^(void){
+      [weakSelf.spinner stopAnimating];
+      
+      //if download failed
+      if (!img) {
+        weakSelf.noImageLoadedView.hidden = NO;
+        return;
+      }
+
+      weakSelf.title = weakSelf.photoInfo.title;
+      
+      weakSelf.imageView = [[UIImageView alloc] initWithImage:img];
+      weakSelf.imageView.frame = CGRectMake(0, 0, img.size.width, img.size.height);
+      
+      [weakSelf.scrollView addSubview:weakSelf.imageView];
+      weakSelf.scrollView.contentSize = weakSelf.imageView.bounds.size;
+    });
+  });
+}
+
+
+- (void)centerScrollViewContents {
+  CGPoint imageViewCenter = self.imageView.center;
+  //we image size after scroll's view transform - thus we take the frame
+  CGSize imageViewSize = CGRectStandardize(self.imageView.frame).size;
+  CGSize scrollViewSize = CGRectStandardize(self.scrollView.bounds).size;
+  
+  CGPoint scrollViewCenter = [self scrollViewCenter];
+  
+  //touch the center only if scrollView exceeds the image
+  
+  if (imageViewSize.width <= scrollViewSize.width) {
+    imageViewCenter.x = scrollViewCenter.x;
+  }
+  
+  if (imageViewSize.height <= scrollViewSize.height) {
+    imageViewCenter.y = scrollViewCenter.y;
+  }
+  
+  //this will put the image in the center and add the padding as needed
+  self.imageView.center = imageViewCenter;
+}
+
+
+- (void)aspectFitImage {
+  CGSize scrollViewSize = CGRectStandardize(self.scrollView.bounds).size;
+  CGSize imageSize = CGRectStandardize(self.imageView.bounds).size;
+  
+  CGFloat scaleWidth = scrollViewSize.width / imageSize.width;
+  CGFloat scaleHeight = scrollViewSize.height / imageSize.height;
+
+  //set the zoom scale so all the image fits the screen
+  self.scrollView.minimumZoomScale = MIN(scaleWidth, scaleHeight);
+  self.scrollView.zoomScale = self.scrollView.minimumZoomScale;
+  self.scrollView.maximumZoomScale = 3;
+  
+  //start zoomed out and centered
+  [self.scrollView setZoomScale:self.scrollView.minimumZoomScale animated:NO];
+  
+
+  self.imageView.center = [self scrollViewCenter];
+}
+
+
+//return the scroll view center
+- (CGPoint)scrollViewCenter {
+  CGSize scrollViewSize = CGRectStandardize(self.scrollView.bounds).size;
+  UIEdgeInsets contentInset = self.scrollView.contentInset;
+  CGFloat centerX = (scrollViewSize.width - contentInset.left - contentInset.right)/2.0;
+  CGFloat centerY = (scrollViewSize.height - contentInset.top - contentInset.bottom)/2.0;
+  
+  return CGPointMake(centerX, centerY);
+}
+
+
+#pragma mark - for debug
+NSString *StrCGPoint(CGPoint p) {
```

Comment:
Good idea. This is perfect for category NSString+CGPoint or NSString+CGDataTypesFactory.